api/enki.py

```diff
@@ -539,226 +479,253 @@ def queue_response(self, status_code, headers={}, content=None):
-        # not mentioned in the results.
-        for identifier_string in identifier_strings:
-            if identifier_string not in seen_identifiers:
-                identifier, ignore = Identifier.for_foreign_id(
-                    self._db, api.ENKI_ID, identifier_string
-                )
-                result = CoverageFailure(
-                    identifier, "Book not found in Enki", data_source=self.output_source, transient=True
-                )
-                batch_results.append(result)
-        return batch_results
-
-    def process_item(self, identifier):
-        results = self.process_batch([identifier])
-        return results[0]
-
-    def handle_success(self, identifier):
-        return self.set_presentation_ready(identifier)
+    log = logging.getLogger("Enki Bibliographic Parser")
 
-class BibliographicParser(object):
+    # Convert the English names of languages given in the Enki API to
+    # the codes we use internally.
+    LANGUAGE_CODES = {
+        "English": u"eng",
+        "French" : u"fre",
+        "Spanish": u"spa",
+    }
 
-    """Helper function to parse JSON"""
     def process_all(self, json_data):
-        data = json.loads(json_data)
-        returned_titles = data["result"]["titles"]
-	titles = returned_titles
+        if isinstance(json_data, basestring):
+            json_data = json.loads(json_data)
+        returned_titles = json_data.get("result", {}).get("titles", [])
 	for book in returned_titles:
-	    data = self.process_one(book)
+            data = self.extract_bibliographic(book)
             if data:
                 yield data
-    log = logging.getLogger("Enki Bibliographic Parser")
-    log.setLevel(logging.DEBUG)
-
-    def __init__(self, include_availability=True, include_bibliographic=True):
-        self.include_availability = include_availability
-        self.include_bibliographic = include_bibliographic
 
     def extract_bibliographic(self, element):
+        """Extract Metadata and CirculationData from a dictionary
+        of information from Enki.
+
+        :return: A Metadata with attached CirculationData.
+        """
+        # TODO: it's not clear what these are or whether we'd find them
+        # useful:
+        #  dateSaved
+        #  length
+        #  publishDate
+        primary_identifier = IdentifierData(EnkiAPI.ENKI_ID, element["id"])
+
         identifiers = []
-        contributors = []
         identifiers.append(IdentifierData(Identifier.ISBN, element["isbn"]))
-        sort_name = element["author"]
-        if not sort_name:
-            sort_name = Edition.UNKNOWN_AUTHOR
+
+        contributors = []
+        sort_name = element.get("author", None) or Edition.UNKNOWN_AUTHOR
         contributors.append(ContributorData(sort_name=sort_name))
-        primary_identifier = IdentifierData(EnkiAPI.ENKI_ID, element["id"])
-        image_url = element["large_image"]
-        thumbnail_url = element["large_image"]
```

Comment:
This error meant the thumbnail and large image were always the same.

Reply:
Update: I did some spot checks and I _think_ it's better to treat the small_image as the thumbnail. I'm not 100% sure but I don't trust 'large_image' to actually be pretty small.